File: PHP/496. Next Greater Element I.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    /**
     * @param  int[]  $nums1
     * @param  int[]  $nums2
     * @return int[]
     */
    public function nextGreaterElement(array $nums1, array $nums2): array
    {
        $map = [];
        $stack = new SplStack;
        for ($i = count($nums2) - 1; $i >= 0; $i--) {
            while (! $stack->isEmpty() && $nums2[$i] >= $stack->top()) {
                $stack->pop();
            }

            $map[$nums2[$i]] = $stack->isEmpty() ? -1 : $stack->top();
            $stack->push($nums2[$i]);
        }

        $elements = [];
        for ($i = 0; $i < count($nums1); $i++) {
            $elements[$i] = $map[$nums1[$i]];
        }

        return $elements;
    }
}

File: Go/496. Next Greater Element I.go
package main

//goland:noinspection GoUnusedFunction
func nextGreaterElement(nums1 []int, nums2 []int) []int {
	elementMap := make(map[int]int)
	stack := make([]int, 0)
	for i := len(nums2) - 1; i >= 0; i-- {
		for len(stack) > 0 && nums2[i] >= stack[len(stack)-1] {
			stack = stack[:len(stack)-1]
		}

		length := len(stack)
		if length == 0 {
			elementMap[nums2[i]] = -1
		} else {
			elementMap[nums2[i]] = stack[length-1]
		}

		stack = append(stack, nums2[i])
	}

	length := len(nums1)
	elements := make([]int, length)
	for i := 0; i < length; i++ {
		elements[i] = elementMap[nums1[i]]
	}
	return elements
}
